src/mappers/country/country_mapper_creator.cpp

```diff
@@ -39,46 +41,137 @@ std::map<std::string, std::string> ImportMappingRules(std::string_view country_m
+          this_->used_hoi4_tags.emplace(mappingRule->second).second)
       {
-         country_mappings.emplace(country.GetNumber(), rule->second);
-         used_hoi4_tags.emplace(rule->second);
+         this_->country_mappings.emplace(country.GetNumber(), mappingRule->second);
+         return true;
       }
-      else
+      return false;
+   }
+
+   // Countries should always be deferred.
+   static bool DeferCountryAlways(CountryMappingCreator* this_, const vic3::Country& country)
+   {
+      this_->deferred_map_countries.push_back(&country);
+      return true;
+   }
+
+   // Countries should be added via their vic3 tag.
+   static bool AddCountryWithVicId(CountryMappingCreator* this_, const vic3::Country& country)
+   {
+      const auto vic3_tag = country.GetTag();
+      if (vic3_tag.length() == 3 && this_->used_hoi4_tags.emplace(vic3_tag).second)
+      {
+         this_->country_mappings.emplace(country.GetNumber(), vic3_tag);
+         return true;
+      }
+      return false;
+   }
+
+   // Countries should be added with Znn naming scheme. This always succeeds.
+   static bool AddCountryWithZ(CountryMappingCreator* this_, const vic3::Country& country)
+   {
+      std::string possible_hoi4_tag;
+      do
       {
-         std::string possible_hoi4_tag = fmt::format("{}{:0>2}", tag_prefix, tag_suffix);
-         while (used_hoi4_tags.contains(possible_hoi4_tag))
+         possible_hoi4_tag = fmt::format("{}{:0>2}", this_->tag_prefix, this_->tag_suffix);
+         ++this_->tag_suffix;
+         if (this_->tag_suffix > 99)
          {
-            ++tag_suffix;
-            if (tag_suffix > 99)
-            {
-               tag_suffix = 0;
-               --tag_prefix;
-            }
-            possible_hoi4_tag = fmt::format("{}{:0>2}", tag_prefix, tag_suffix);
+            this_->tag_suffix = 0;
+            --this_->tag_prefix;
          }
-         country_mappings.emplace(country.GetNumber(), possible_hoi4_tag);
-         used_hoi4_tags.emplace(possible_hoi4_tag);
+      } while (this_->used_hoi4_tags.contains(possible_hoi4_tag));
+      this_->country_mappings.emplace(country.GetNumber(), possible_hoi4_tag);
+      this_->used_hoi4_tags.emplace(possible_hoi4_tag);
+      return true;
+   }
+
+   // Attempt to name (or defer naming of) a country. Only the first successful strategy will be used.
+   void ExecuteStrategiesForCountry(const vic3::Country& country, const std::vector<CountryStrategyFn*> strategies)
+   {
+      for (decltype(strategies)::const_iterator strategy = strategies.begin(); strategy != strategies.end(); ++strategy)
+      {
+         if ((**strategy)(this, country))
+         {
+            return;
+         }
+      }
+   }
+
+  public:
+   CountryMappingCreator(std::string_view country_mappings_file):
+       country_mapping_rules(ImportMappingRules(country_mappings_file))
+   {
+   }
+
+   // Where the logic happens
+   std::map<int, std::string>&& AssignTags(auto countries)
```

Comment:
Returning by r-value seems like it may be overkill. It might be worth double-checking with some sample code on godbolt.org to see if it really eliminates a copy. Feel free to disregard if too much work though. :-D

Reply:
I think you're right that it creates a copy because CountryMapper takes a map&, not a map&&.
However, I will keep this in because profiling revealed that we spend a *lot* of time copying data, and a map of all countries seems like a Big Struct, and this is easily identifiable as a situation where we *could* move instead of copying.